### ISSUE TITLE
RFC: Versioned .js.flow files

### ISSUE THREAD

**Issue by willmruzek** [facebook/flow#4917]
# The Problem

Imagine you have a shared library of (React) components shared across your company's projects. 

This library is on its own version of flow, while all consumers are also on different versions of flow.

Sometimes this results in different flow errors in each project. For example, a comment suppression may be needed and valid in the library, but the consuming app's version of flow throws an `Unused suppression` error. Another case might be the use of opaque types in the library, but the app's flow doesn't support them yet. 

Another side effect is that it's difficult to make major upgrades (e.g. 0.53.x regarding React).

# Proposed Solution

Currently flow consumes *.js.flow files found in a published (public or private) package.

To solve the problem, flow could also consume version-tagged files.

For example, type definitions intended for 0.52 would be available with a *.js.flow-0.52.x file

Note: Generating the different versions of types would be a separate issue

# Why not...

## Force consumers to upgrade

This is a reasonable solution with minor changes between versions, but quickly becomes unreasonable with larger changes (again, 0.53.x regarding React).

## Keep with the lowest common denominator

Flow keeps getting better with each version. It would be a shame to miss out on all the improvements. This would greatly slow the adoption of a new version.

## Submit types to flow-typed

In many cases, companies do not want to release or open source their code. There needs to be a way to share (versioned) types privately.

## Build and maintain a private flow-typed repository

This would be a viable option. The flow-typed CLI already contains the logic to resolve libdefs with different versions of flow. However, the company would need to set up and maintain their own repository and tools. It's possible the resources aren't there to do this, or it's not worth it for one library.

I'm happy to submit a PR, but I wanted to refine the idea first with the community before starting.

**Comment by jedwards1211** [comment]
@mruzekw since `flow` itself and tools like `flow-upgrade` rely on finding files ending in `.flow`, maybe it would be better to for the defs for `src/foo.js` to live in `src/flow-typed/<flow version>/foo.js.flow` instead.

**Comment by mruzekw** [comment]
@jedwards1211 Thanks for your input.

I thought .js.flow files were usually automatically copied from the original source. Why would flow-upgrade care about derivative files?

**Comment by jedwards1211** [comment]
@mruzekw what do you mean automatically copied?  Are you talking about `gen-flow-files` (which last I checked, doesn't always work?)  Or does `cp file.js file.js.flow` typically work well enough?  I've been hand-writing `.js.flow` files with only typedefs in them...

I can say though that there is code in `flow-upgrade` that looks for files with `.flow` extensions.

**Comment by jedwards1211** [comment]
@mruzekw regardless, for flow itself to search for the corresponing `.js.flow-*` files might be less efficient than looking in a `flow-typed/<flow version>` subdirectory.

**Comment by mruzekw** [comment]
@jedwards1211 The issue I have with `flow-typed` is that these types (of private projects) have nothing to do with flow-typed.

BTW, we generate our .flow.js files during the build with https://github.com/AgentME/flow-copy-source

I don't know what other projects do.

**Comment by asolove** [comment]
I’m not sure about this exact proposal, but I definitely agree with the overall pain point of managing libraries across different versions of flow.

I’ll add another issue we have with flow-typed: we can write the files by hand, but there isn’t (to my knowledge) a way to take the definitions and actually check them against the implementation code. This makes it tricky to feel confident in their correctness. It would be great to have a tool to check declarations against an implementing file. 

All this said, I don’t think we can expect the flow core team to solve this problem. They are focused on facebooks use case, where I think they have a largely monolithic codebase and a single version of flow in use at one time. So I think the community is going to have to provide the labor to find a solution to this issue.

**Comment by mruzekw** [comment]
@asolove Do you have any ideas how we could solve the issue of private libdefs + different versions of flow without altering flow itself?

**Comment by asolove** [comment]
That’s all I mean, we’re going to have to write the pr ourselves. FB have been clear the core team aren’t working on third party only issues. I’d be happy to help if we had a good approach and had an approach that the core team approved.

**Comment by jedwards1211** [comment]
Weird, I never realized plain old source files would work as `.js.flow` files.  I thought they had to have declare statements in them

**Comment by jedwards1211** [comment]
@mruzekw If you don't want to modify flow itself, another option would be to make a PR to `flow-typed`: have it look for a top-level `flow-typed` directory in the package in question (instead of the corresponding directory in the `flow-typed` repo), and if one exists, copy the defs for the correct flow version from it.

I don't know about you guys, but I've never used OCaml so making a PR to flow itself would take some digging, but making `flow-typed` handle the above would probably be a breeze.

As far as the idea of making flow directly support defs for multiple versions though, we still haven't agreed on a file or directory naming convention.  Here are two other proposals for that:
1. `somedir/foo.js` -> `somedir/foo.flow-v0.28.x-.js.flow`
2. `somedir/foo.js` -> `somedir/flow-defs/flow-v0.28.x-/foo.js.flow`

**Comment by jedwards1211** [comment]
@asolove @mruzekw I just make a PR to `flow-typed` to install defs stored in a `flow-typed` directory in a published package.  Check it out!

**Comment by mruzekw** [comment]
@jedwards1211 Correct me if I'm wrong, but this doesn't solve the need for privately versioned flow defs.

**Comment by jedwards1211** [comment]
@mruzekw yes it does, you publish your private package with a `flow-typed` directory under its root containing subdirectories with defs for various flow versions.

**Comment by mruzekw** [comment]
@jedwards1211 Oh, I see. So I'd need to run `flow-typed install` or `flow-typed update`?

**Comment by jedwards1211** [comment]
@mruzekw first you'd need to make sure you've run `npm install` or `yarn` to make sure the packages are present.  Then you run `flow-typed install` or `flow-typed update`.

**Comment by mruzekw** [comment]
@jedwards1211 Interesting! Continuing further questions on the PR.

**Comment by SamChou19815** [comment]
.js.flow files are like .d.ts files. They shouldn't be version aware.
